File: src/voxel_array.rs
use std::ptr;

use crate::error::Result;
use crate::ffi;
use crate::handle::ObjectHandle;
use crate::mesh::Mesh;
use crate::object::Object;
use crate::types::{BoundingBox, VoxelArrayInfo, VoxelIndexExtent};
use crate::util::{parse_json, required_handle};

#[derive(Debug, Clone)]
pub struct VoxelArray {
    handle: ObjectHandle,
}

impl VoxelArray {
    pub(crate) fn from_handle(handle: ObjectHandle) -> Self {
        Self { handle }
    }

    pub fn new(
        voxel_indices: &[[i32; 4]],
        bounding_box: BoundingBox,
        voxel_extent: f32,
    ) -> Result<Self> {
        let flattened = voxel_indices
            .iter()
            .flat_map(|index| index.iter().copied())
            .collect::<Vec<_>>();
        let mut out_voxel_array = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        let status = unsafe {
            ffi::mdl_voxel_array_new_with_indices(
                flattened.as_ptr(),
                voxel_indices.len() as u64,
                bounding_box.min[0],
                bounding_box.min[1],
                bounding_box.min[2],
                bounding_box.max[0],
                bounding_box.max[1],
                bounding_box.max[2],
                voxel_extent,
                &mut out_voxel_array,
                &mut out_error,
            )
        };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(
            out_voxel_array,
            "MDLVoxelArray",
        )?))
    }

    pub fn info(&self) -> Result<VoxelArrayInfo> {
        parse_json(
            unsafe { ffi::mdl_voxel_array_info_json(self.handle.as_ptr()) },
            "MDLVoxelArray",
        )
    }

    #[must_use]
    pub fn count(&self) -> usize {
        unsafe { ffi::mdl_voxel_array_count(self.handle.as_ptr()) as usize }
    }

    pub fn set_voxel(&self, index: [i32; 4]) {
        unsafe {
            ffi::mdl_voxel_array_set_voxel(
                self.handle.as_ptr(),
                index[0],
                index[1],
                index[2],
                index[3],
            );
        };
    }

    #[must_use]
    #[allow(clippy::fn_params_excessive_bools)]
    pub fn voxel_exists(
        &self,
        index: [i32; 4],
        allow_any_x: bool,
        allow_any_y: bool,
        allow_any_z: bool,
        allow_any_shell: bool,
    ) -> bool {
        unsafe {
            ffi::mdl_voxel_array_voxel_exists(
                self.handle.as_ptr(),
                index[0],
                index[1],
                index[2],
                index[3],
                i32::from(allow_any_x),
                i32::from(allow_any_y),
                i32::from(allow_any_z),
                i32::from(allow_any_shell),
            ) != 0
        }
    }

    #[must_use]
    pub fn voxel_indices(&self) -> Vec<[i32; 4]> {
        let count = self.count();
        if count == 0 {
            return Vec::new();
        }
        let mut flattened = vec![0_i32; count * 4];
        let written = unsafe {
            ffi::mdl_voxel_array_copy_indices(
                self.handle.as_ptr(),
                flattened.as_mut_ptr(),
                count as u64,
            )
        } as usize;
        flattened.truncate(written * 4);
        flattened
            .chunks_exact(4)
            .map(|chunk| [chunk[0], chunk[1], chunk[2], chunk[3]])
            .collect()
    }

    #[must_use]
    pub fn voxels_within_extent(&self, extent: &VoxelIndexExtent) -> Vec<[i32; 4]> {
        let capacity = self.count();
        if capacity == 0 {
            return Vec::new();
        }
        let mut flattened = vec![0_i32; capacity * 4];
        let written = unsafe {
            ffi::mdl_voxel_array_copy_voxels_within_extent(
                self.handle.as_ptr(),
                extent.minimum_extent[0],
                extent.minimum_extent[1],
                extent.minimum_extent[2],
                extent.minimum_extent[3],
                extent.maximum_extent[0],
                extent.maximum_extent[1],
                extent.maximum_extent[2],
                extent.maximum_extent[3],
                flattened.as_mut_ptr(),
                capacity as u64,
            )
        } as usize;
        flattened.truncate(written * 4);
        flattened
            .chunks_exact(4)
            .map(|chunk| [chunk[0], chunk[1], chunk[2], chunk[3]])
            .collect()
    }

    pub fn union_with(&self, other: &Self) {
        unsafe { ffi::mdl_voxel_array_union(self.handle.as_ptr(), other.handle.as_ptr()) };
    }

    pub fn intersect_with(&self, other: &Self) {
        unsafe { ffi::mdl_voxel_array_intersect(self.handle.as_ptr(), other.handle.as_ptr()) };
    }

    pub fn difference_with(&self, other: &Self) {
        unsafe { ffi::mdl_voxel_array_difference(self.handle.as_ptr(), other.handle.as_ptr()) };
    }

    #[must_use]
    pub fn index_of_spatial_location(&self, location: [f32; 3]) -> [i32; 4] {
        let mut values = [0_i32; 4];
        unsafe {
            ffi::mdl_voxel_array_index_of_spatial_location(
                self.handle.as_ptr(),
                location[0],
                location[1],
                location[2],
                values.as_mut_ptr(),
            );
        };
        values
    }

    #[must_use]
    pub fn spatial_location_of_index(&self, index: [i32; 4]) -> [f32; 3] {
        let mut values = [0.0_f32; 3];
        unsafe {
            ffi::mdl_voxel_array_spatial_location_of_index(
                self.handle.as_ptr(),
                index[0],
                index[1],
                index[2],
                index[3],
                &mut values[0],
                &mut values[1],
                &mut values[2],
            );
        };
        values
    }

    #[must_use]
    pub fn voxel_bounding_box_at_index(&self, index: [i32; 4]) -> BoundingBox {
        let mut min = [0.0_f32; 3];
        let mut max = [0.0_f32; 3];
        unsafe {
            ffi::mdl_voxel_array_voxel_bounding_box_at_index(
                self.handle.as_ptr(),
                index[0],
                index[1],
                index[2],
                index[3],
                &mut min[0],
                &mut min[1],
                &mut min[2],
                &mut max[0],
                &mut max[1],
                &mut max[2],
            );
        };
        BoundingBox { min, max }
    }

    pub fn convert_to_signed_shell_field(&self) {
        unsafe { ffi::mdl_voxel_array_convert_to_signed_shell_field(self.handle.as_ptr()) };
    }

    pub fn set_shell_field_interior_thickness(&self, value: f32) {
        unsafe {
            ffi::mdl_voxel_array_set_shell_field_interior_thickness(self.handle.as_ptr(), value);
        };
    }

    pub fn set_shell_field_exterior_thickness(&self, value: f32) {
        unsafe {
            ffi::mdl_voxel_array_set_shell_field_exterior_thickness(self.handle.as_ptr(), value);
        };
    }

    #[must_use]
    pub fn coarse_mesh(&self) -> Option<Mesh> {
        let ptr = unsafe { ffi::mdl_voxel_array_coarse_mesh(self.handle.as_ptr()) };
        unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(Mesh::from_handle)
    }

    #[must_use]
    pub fn mesh(&self) -> Option<Mesh> {
        let ptr = unsafe { ffi::mdl_voxel_array_mesh(self.handle.as_ptr()) };
        unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(Mesh::from_handle)
    }

    #[must_use]
    pub fn as_object(&self) -> Object {
        Object::from_handle(self.handle.clone())
    }
}

File: swift-bridge/Sources/ModelIOBridge/VoxelArray.swift
import Foundation
import ModelIO
import simd

private func mdl_voxel_index_extent(_ extent: MDLVoxelIndexExtent) -> [String: Any] {
    [
        "minimum_extent": [extent.minimumExtent.x, extent.minimumExtent.y, extent.minimumExtent.z, extent.minimumExtent.w],
        "maximum_extent": [extent.maximumExtent.x, extent.maximumExtent.y, extent.maximumExtent.z, extent.maximumExtent.w],
    ]
}

private func mdl_voxel_array_info(_ voxelArray: MDLVoxelArray) -> [String: Any] {
    [
        "count": voxelArray.count,
        "bounding_box": mdl_bounding_box(voxelArray.boundingBox),
        "voxel_index_extent": mdl_voxel_index_extent(voxelArray.voxelIndexExtent),
        "is_valid_signed_shell_field": voxelArray.isValidSignedShellField,
        "shell_field_interior_thickness": voxelArray.shellFieldInteriorThickness,
        "shell_field_exterior_thickness": voxelArray.shellFieldExteriorThickness,
    ]
}

private func mdl_voxel_indices(_ data: Data?) -> [Int32] {
    guard let data else { return [] }
    return data.withUnsafeBytes { buffer in
        Array(buffer.bindMemory(to: Int32.self))
    }
}

@_cdecl("mdl_voxel_array_new_with_asset")
public func mdl_voxel_array_new_with_asset(
    _ assetHandle: UnsafeMutableRawPointer?,
    _ divisions: Int32,
    _ patchRadius: Float,
    _ outVoxelArray: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let asset = mdl_borrow_object(assetHandle) as? MDLAsset,
              let outVoxelArray
        else {
            throw ModelIOBridgeError.invalidArgument("missing asset or output voxel array pointer")
        }
        outVoxelArray.pointee = mdl_retain(MDLVoxelArray(asset: asset, divisions: divisions, patchRadius: patchRadius))
    }
}

@_cdecl("mdl_voxel_array_new_with_indices")
public func mdl_voxel_array_new_with_indices(
    _ values: UnsafePointer<Int32>?,
    _ count: UInt64,
    _ minX: Float,
    _ minY: Float,
    _ minZ: Float,
    _ maxX: Float,
    _ maxY: Float,
    _ maxZ: Float,
    _ voxelExtent: Float,
    _ outVoxelArray: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let outVoxelArray else {
            throw ModelIOBridgeError.invalidArgument("missing output voxel array pointer")
        }
        let data = mdl_data_from_int32s(values, count: count * 4)
        let boundingBox = MDLAxisAlignedBoundingBox(
            maxBounds: SIMD3<Float>(maxX, maxY, maxZ),
            minBounds: SIMD3<Float>(minX, minY, minZ)
        )
        outVoxelArray.pointee = mdl_retain(MDLVoxelArray(data: data, boundingBox: boundingBox, voxelExtent: voxelExtent))
    }
}

@_cdecl("mdl_voxel_array_info_json")
public func mdl_voxel_array_info_json(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray else { return nil }
    return mdl_string(mdl_json_string(from: mdl_voxel_array_info(voxelArray)) ?? "{}")
}

@_cdecl("mdl_voxel_array_count")
public func mdl_voxel_array_count(_ handle: UnsafeMutableRawPointer?) -> UInt64 {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray else { return 0 }
    return UInt64(voxelArray.count)
}

@_cdecl("mdl_voxel_array_set_voxels_for_mesh")
public func mdl_voxel_array_set_voxels_for_mesh(
    _ handle: UnsafeMutableRawPointer?,
    _ meshHandle: UnsafeMutableRawPointer?,
    _ divisions: Int32,
    _ patchRadius: Float
) {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray,
          let mesh = mdl_borrow_object(meshHandle) as? MDLMesh
    else {
        return
    }
    voxelArray.setVoxelsFor(mesh, divisions: divisions, patchRadius: patchRadius)
}

@_cdecl("mdl_voxel_array_set_voxel")
public func mdl_voxel_array_set_voxel(
    _ handle: UnsafeMutableRawPointer?,
    _ x: Int32,
    _ y: Int32,
    _ z: Int32,
    _ shell: Int32
) {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray else { return }
    voxelArray.setVoxelAtIndex(vector_int4(x, y, z, shell))
}

@_cdecl("mdl_voxel_array_voxel_exists")
public func mdl_voxel_array_voxel_exists(
    _ handle: UnsafeMutableRawPointer?,
    _ x: Int32,
    _ y: Int32,
    _ z: Int32,
    _ shell: Int32,
    _ allowAnyX: Int32,
    _ allowAnyY: Int32,
    _ allowAnyZ: Int32,
    _ allowAnyShell: Int32
) -> Int32 {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray else { return 0 }
    return voxelArray.voxelExists(
        atIndex: vector_int4(x, y, z, shell),
        allowAnyX: allowAnyX != 0,
        allowAnyY: allowAnyY != 0,
        allowAnyZ: allowAnyZ != 0,
        allowAnyShell: allowAnyShell != 0
    ) ? 1 : 0
}

@_cdecl("mdl_voxel_array_copy_indices")
public func mdl_voxel_array_copy_indices(
    _ handle: UnsafeMutableRawPointer?,
    _ outValues: UnsafeMutablePointer<Int32>?,
    _ capacityIndices: UInt64
) -> UInt64 {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray,
          let outValues
    else {
        return 0
    }
    let values = mdl_voxel_indices(voxelArray.voxelIndices())
    let totalIndices = values.count / 4
    let copyCount = min(Int(capacityIndices), totalIndices)
    guard copyCount > 0 else { return 0 }
    let flattened = Array(values.prefix(copyCount * 4))
    flattened.withUnsafeBufferPointer { buffer in
        outValues.initialize(from: buffer.baseAddress!, count: flattened.count)
    }
    return UInt64(copyCount)
}

@_cdecl("mdl_voxel_array_copy_voxels_within_extent")
public func mdl_voxel_array_copy_voxels_within_extent(
    _ handle: UnsafeMutableRawPointer?,
    _ minX: Int32,
    _ minY: Int32,
    _ minZ: Int32,
    _ minShell: Int32,
    _ maxX: Int32,
    _ maxY: Int32,
    _ maxZ: Int32,
    _ maxShell: Int32,
    _ outValues: UnsafeMutablePointer<Int32>?,
    _ capacityIndices: UInt64
) -> UInt64 {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray,
          let outValues
    else {
        return 0
    }
    let extent = MDLVoxelIndexExtent(
        minimumExtent: vector_int4(minX, minY, minZ, minShell),
        maximumExtent: vector_int4(maxX, maxY, maxZ, maxShell)
    )
    let values = mdl_voxel_indices(voxelArray.voxels(within: extent))
    let totalIndices = values.count / 4
    let copyCount = min(Int(capacityIndices), totalIndices)
    guard copyCount > 0 else { return 0 }
    let flattened = Array(values.prefix(copyCount * 4))
    flattened.withUnsafeBufferPointer { buffer in
        outValues.initialize(from: buffer.baseAddress!, count: flattened.count)
    }
    return UInt64(copyCount)
}

@_cdecl("mdl_voxel_array_union")
public func mdl_voxel_array_union(_ handle: UnsafeMutableRawPointer?, _ otherHandle: UnsafeMutableRawPointer?) {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray,
          let other = mdl_borrow_object(otherHandle) as? MDLVoxelArray
    else {
        return
    }
    voxelArray.union(with: other)
}

@_cdecl("mdl_voxel_array_intersect")
public func mdl_voxel_array_intersect(_ handle: UnsafeMutableRawPointer?, _ otherHandle: UnsafeMutableRawPointer?) {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray,
          let other = mdl_borrow_object(otherHandle) as? MDLVoxelArray
    else {
        return
    }
    voxelArray.intersect(with: other)
}

@_cdecl("mdl_voxel_array_difference")
public func mdl_voxel_array_difference(_ handle: UnsafeMutableRawPointer?, _ otherHandle: UnsafeMutableRawPointer?) {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray,
          let other = mdl_borrow_object(otherHandle) as? MDLVoxelArray
    else {
        return
    }
    voxelArray.difference(with: other)
}

@_cdecl("mdl_voxel_array_index_of_spatial_location")
public func mdl_voxel_array_index_of_spatial_location(
    _ handle: UnsafeMutableRawPointer?,
    _ x: Float,
    _ y: Float,
    _ z: Float,
    _ outValues: UnsafeMutablePointer<Int32>?
) {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray,
          let outValues
    else {
        return
    }
    let index = voxelArray.index(ofSpatialLocation: SIMD3<Float>(x, y, z))
    outValues[0] = index.x
    outValues[1] = index.y
    outValues[2] = index.z
    outValues[3] = index.w
}

@_cdecl("mdl_voxel_array_spatial_location_of_index")
public func mdl_voxel_array_spatial_location_of_index(
    _ handle: UnsafeMutableRawPointer?,
    _ x: Int32,
    _ y: Int32,
    _ z: Int32,
    _ shell: Int32,
    _ outX: UnsafeMutablePointer<Float>?,
    _ outY: UnsafeMutablePointer<Float>?,
    _ outZ: UnsafeMutablePointer<Float>?
) {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray else { return }
    let location = voxelArray.spatialLocation(ofIndex: vector_int4(x, y, z, shell))
    outX?.pointee = location.x
    outY?.pointee = location.y
    outZ?.pointee = location.z
}

@_cdecl("mdl_voxel_array_voxel_bounding_box_at_index")
public func mdl_voxel_array_voxel_bounding_box_at_index(
    _ handle: UnsafeMutableRawPointer?,
    _ x: Int32,
    _ y: Int32,
    _ z: Int32,
    _ shell: Int32,
    _ outMinX: UnsafeMutablePointer<Float>?,
    _ outMinY: UnsafeMutablePointer<Float>?,
    _ outMinZ: UnsafeMutablePointer<Float>?,
    _ outMaxX: UnsafeMutablePointer<Float>?,
    _ outMaxY: UnsafeMutablePointer<Float>?,
    _ outMaxZ: UnsafeMutablePointer<Float>?
) {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray else { return }
    let boundingBox = voxelArray.voxelBoundingBox(atIndex: vector_int4(x, y, z, shell))
    outMinX?.pointee = boundingBox.minBounds.x
    outMinY?.pointee = boundingBox.minBounds.y
    outMinZ?.pointee = boundingBox.minBounds.z
    outMaxX?.pointee = boundingBox.maxBounds.x
    outMaxY?.pointee = boundingBox.maxBounds.y
    outMaxZ?.pointee = boundingBox.maxBounds.z
}

@_cdecl("mdl_voxel_array_convert_to_signed_shell_field")
public func mdl_voxel_array_convert_to_signed_shell_field(_ handle: UnsafeMutableRawPointer?) {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray else { return }
    voxelArray.convertToSignedShellField()
}

@_cdecl("mdl_voxel_array_set_shell_field_interior_thickness")
public func mdl_voxel_array_set_shell_field_interior_thickness(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray else { return }
    voxelArray.shellFieldInteriorThickness = value
}

@_cdecl("mdl_voxel_array_set_shell_field_exterior_thickness")
public func mdl_voxel_array_set_shell_field_exterior_thickness(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray else { return }
    voxelArray.shellFieldExteriorThickness = value
}

@_cdecl("mdl_voxel_array_coarse_mesh")
public func mdl_voxel_array_coarse_mesh(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    mdl_voxel_array_coarse_mesh_with_allocator(handle, nil)
}

@_cdecl("mdl_voxel_array_coarse_mesh_with_allocator")
public func mdl_voxel_array_coarse_mesh_with_allocator(
    _ handle: UnsafeMutableRawPointer?,
    _ allocatorHandle: UnsafeMutableRawPointer?
) -> UnsafeMutableRawPointer? {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray else { return nil }
    let allocator = mdl_borrow_object(allocatorHandle) as? any MDLMeshBufferAllocator
    guard let mesh = voxelArray.coarseMesh(using: allocator) else { return nil }
    return mdl_retain(mesh)
}

@_cdecl("mdl_voxel_array_mesh")
public func mdl_voxel_array_mesh(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    mdl_voxel_array_mesh_with_allocator(handle, nil)
}

@_cdecl("mdl_voxel_array_mesh_with_allocator")
public func mdl_voxel_array_mesh_with_allocator(
    _ handle: UnsafeMutableRawPointer?,
    _ allocatorHandle: UnsafeMutableRawPointer?
) -> UnsafeMutableRawPointer? {
    guard let voxelArray = mdl_borrow_object(handle) as? MDLVoxelArray else { return nil }
    let allocator = mdl_borrow_object(allocatorHandle) as? any MDLMeshBufferAllocator
    guard let mesh = voxelArray.mesh(using: allocator) else { return nil }
    return mdl_retain(mesh)
}
